### PR TITLE
Develop hybrid action staging

### PR DESCRIPTION
**Pull request by andrewcoh**
### Proposed change(s)

Hybrid Actions staging branch

To merge in before it is ready to merge into master:

- ~~ActionTuple~~
- ~~ActionModel~~
- ~~C#~~

TODO:

- ~~Merge master~~ Addressing clipped actions was non-trivial. Please look at how ActionInfo and clipping in the torch policy has changed.
- ~~Fix GAIL simple RL (after C#)~~ (GAIL tests passed but still using demonstrations with old proto fields)
- (may come later or separate PR) Update the demo files
- ~~Docs~~
- Env (in separate PR)

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/policy/policy.py [hunk 1]
```diff
@@ -142,14 +132,18 @@ def get_action(
...
+    def check_nan_action(action: Optional[ActionTuple]) -> None:
         # Fast NaN check on the action
         # See https://stackoverflow.com/questions/6736590/fast-check-for-nan-in-numpy for background.
         if action is not None:
-            d = np.sum(action)
+            d = np.sum(action.continuous)
+            has_nan = np.isnan(d)
+            if has_nan:
+                raise RuntimeError("Continuous NaN action detected.")
+            d = np.sum(action.discrete)
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/policy/policy.py:143]: action.discrete is int32's right? I don't think we need to check them for NaNs

Comment (chriselion) [reply ml-agents/mlagents/trainers/policy/policy.py:143]: (OK to do after merging to master)

**Review by chriselion — APPROVED** [review]

File: docs/Learning-Environment-Design-Agents.md [hunk 2]
```diff
@@ -653,26 +639,26 @@ nothing"_ or _"change weapon"_ for his next decision (since action index 1 and 2
...
   branches.
 - You cannot mask all the actions of a branch.
 - You cannot mask actions in continuous control.
 
 ### Actions Summary & Best Practices
 
-- Actions can either use `Discrete` or `Continuous` spaces.
-- When using `Discrete` it is possible to assign multiple action branches, and
-  to mask certain actions.
+- Agents can either use `Discrete` or `Continuous` actions.
```
Comment (ervteng) [inline docs/Learning-Environment-Design-Agents.md:654]:
```suggestion
- Agents can use `Discrete` and/or `Continuous` actions.
```

File: docs/Python-API.md [hunk 3]
```diff
@@ -230,33 +226,27 @@ A `TerminalStep` has the following fields:
...
-  - In continuous action space it is the number of floats that constitute the
-    action.
-  - In discrete action space (same as multi-discrete) it corresponds to the
-    number of branches (the number of independent actions)
-- `discrete_action_branches` is a Tuple of int only for discrete action space.
-  Each int corresponds to the number of different options for each branch of the
-  action. For example : In a game direction input (no movement, left, right) and
+- `action_spec` is an `ActionSpec` namedtuple that defines the number and types
+  of actions for the Agent.
+
```
Comment (ervteng) [inline docs/Python-API.md:237]:
```suggestion
#### ActionSpec
```

**Review by dongruoping — APPROVED** [review]
Changelog hasn't been updated yet.

**Comment by andrewcoh** [comment]
> Changelog hasn't been updated yet.

I'll update in a subsequent PR.

**Review by ervteng — APPROVED** [review]
:shipit:

**Pull request merged**